File: ios/MyTwitter/TweetUrlParser.swift
import Foundation

enum TweetUrlParser {
    private static let statusIdRegex = try! NSRegularExpression(
        pattern: #"/(?:i/)?(?:web/)?status(?:es)?/(\d+)"#,
        options: .caseInsensitive
    )

    private static let xHosts: Set<String> = [
        "x.com",
        "www.x.com",
        "mobile.x.com",
        "twitter.com",
        "www.twitter.com",
        "mobile.twitter.com",
    ]

    static func isXStatusOrTco(_ url: URL) -> Bool {
        guard let host = url.host?.lowercased() else { return false }
        if host == "t.co" { return true }
        guard xHosts.contains(host) else { return false }
        let path = url.path
        let range = NSRange(path.startIndex..<path.endIndex, in: path)
        return statusIdRegex.firstMatch(in: path, options: [], range: range) != nil
    }

    static func extractStatusId(_ url: URL) -> String? {
        guard let host = url.host?.lowercased(), xHosts.contains(host) else { return nil }
        let path = url.path
        let range = NSRange(path.startIndex..<path.endIndex, in: path)
        guard let match = statusIdRegex.firstMatch(in: path, options: [], range: range),
              match.numberOfRanges > 1,
              let idRange = Range(match.range(at: 1), in: path)
        else { return nil }
        return String(path[idRange])
    }
}

File: android/app/src/main/java/org/evergreenlabs/mytwitter/TweetUrlParser.kt
package org.evergreenlabs.mytwitter

import android.net.Uri

object TweetUrlParser {
    private val statusIdRegex =
        Regex("""/(?:i/)?(?:web/)?status(?:es)?/(\d+)""", RegexOption.IGNORE_CASE)

    fun isXStatusOrTco(uri: Uri): Boolean {
        val host = uri.host?.lowercase() ?: return false
        if (host == "t.co") return true
        if (host !in X_HOSTS) return false
        return statusIdRegex.containsMatchIn(uri.path ?: "")
    }

    fun extractStatusId(uri: Uri): String? {
        val host = uri.host?.lowercase() ?: return null
        if (host == "t.co") return null
        if (host !in X_HOSTS) return null
        return statusIdRegex.find(uri.path ?: "")?.groupValues?.getOrNull(1)
    }

    private val X_HOSTS = setOf(
        "x.com",
        "www.x.com",
        "mobile.x.com",
        "twitter.com",
        "www.twitter.com",
        "mobile.twitter.com",
    )
}
